Add first runable version

taloric opened the pull request:
- DONE: 
  - refactor command line args parse ( conflict in last version)
  - add env/commandline args parse option
  - add logs
  - add election logic (add plugins k8s/etcd/redis)
  - add server scheduler code logic
  - add prometheus liveness check & locker liveness check

- TODO:
  - unit test
  - integration test
  - extra functions...

Commit:
support k8s and etcd Elections.
support adapter receive prometheus remote write and adapter remote write to remote storage.

Commit:
support k8s and etcd Elections.
support prometheus remote write to adapter and adapter remote write.

Commit:
push go.mod

taloric committed:
init base version for dryrun

taloric commented:
@chenmudu please review

chenmudu commented on pkg/utils/process.go:
```diff
@@ -0,0 +1,46 @@
+	return filepath.Base(path)
+}
+
+func GetLocalIP() string {
```
If this code is not being used, consider removing it.

chenmudu commented on pkg/service/scheduler.go:
```diff
@@ -0,0 +1,50 @@
+	for range ticker.C {
+		elapsed := time.Since(time.Unix(0, svc.lastReceiveTime))
+		if svc.elector.IsLeader() {
+			if elapsed > svc.conf.PrometheusTimeout {
```
Maybe we can increase the gap here to prevent frequent switching due to network fluctuations.
Like: elapsed > svc.conf.PrometheusTimeout + magicTime.

chenmudu replied:
The magic number can be user supplied, but there must be a default value, 10 seconds is recommended.

chenmudu replied:
In addition, we need to tell the user in the document that PrometheusTimeout == scrape_interval(prometheus config).

taloric replied:
> The magic number can be user supplied

I think... let's consider such 2 scenes below: 
1. we offer `prometheus timeout ` config, then user make it equals `scrape_interval`, we add 10secs as `magic timeout number` for detect prometheus time out. When I find out `prometheus timeout` happened, I can add 10secs for `prometheus timeout ` config. That's easy to understand. 

2. we offer `promtheus timeout` and `magic number` config, then user make `promtheus timeout` equals `scrape_interval` and update `magic timeout number`  as 10secs or more. Here brings an confusing issue: as a user, why don't I make `prometheus timeout` just equals `scrape_interval` +10secs? why I need to make 2 configs for one feature? Which one of those two configs is available?

chenmudu replied:
> 2. we offer `promtheus timeout` and `magic number` config, then user make `promtheus timeout` equals `scrape_interval` and update `magic timeout number`  as 10secs or more. Here brings an confusing issue: as a user, why don't I make `prometheus timeout` just equals `scrape_interval` +10secs? why I need to make 2 configs for one feature? Which one of those two configs is available?

The reason for this is that users only care about  `scrape_interval`  and know what `scrape_interval` means, not what  `promtheus timeout`  is defined internally.

taloric committed:
Fix bugs and remove some configs.

- complete k8s & redis elector verification
- integration test with real remote_write & push to next remote_write

chenmudu commented:
LGTM.
